### PR TITLE
Added IRequestDelegateFactory

### PR DESCRIPTION
**Pull request by davidfowl**
- This will allow us to generate a unique request delegate based on the action descriptor at early in the pipeline.

PS: This is mostly plumbing code.

### CODE REVIEW

**Comment by davidfowl** [comment]
I'm torn on whether this should be exposed, but we can always do that later I guess.

File: src/Mvc/Mvc.Core/src/Infrastructure/ControllerActionInvokerCache.cs [hunk 1]
```diff
@@ -43,30 +36,16 @@ internal class ControllerActionInvokerCache
...
         }
 
-        private InnerCache CurrentCache
-        {
-            get
-            {
-                var current = _currentCache;
-                var actionDescriptors = _collectionProvider.ActionDescriptors;
-
-                if (current == null || current.Version != actionDescriptors.Version)
```
Comment (halter73) [inline src/Mvc/Mvc.Core/src/Infrastructure/ControllerActionInvokerCache.cs:53]: Don't we still need the IActionDescriptorCollectionProvider so we can the current ActionDescriptor.Version matches ActionDescriptorCollection.Version and invalidate the CacheEntry?

File: src/Mvc/Mvc.Core/src/Routing/IRequestDelegateFactory.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    /// <summary>
+    /// Internal interfaces that allows us to optimize the request execution path based on ActionDescriptor
+    /// </summary>
+    internal interface IRequestDelegateFactory
```
Comment (halter73) [inline src/Mvc/Mvc.Core/src/Routing/IRequestDelegateFactory.cs:13]: It is a little odd to merge an abstraction without any implementations. I know there's https://github.com/dotnet/aspnetcore/blob/e9d16525cf3a832631a4e29ce826290adae873a9/src/Mvc/Mvc.Core/src/Routing/ControllerRequestDelegateFactory.cs ofc. Maybe we should wait for that or some other implementation of IRequestDelegateFactory to be merged before taking this.

Comment (davidfowl) [reply src/Mvc/Mvc.Core/src/Routing/IRequestDelegateFactory.cs:13]: Agreed. I'll merge both PRs

**Comment by davidfowl** [comment]
Closed in favor of https://github.com/dotnet/aspnetcore/pull/27773

**Pull request closed without merging**